webapp/channels/src/components/analytics/doughnut_chart.tsx

```diff
@@ -16,30 +17,56 @@ type Props = {
 const DoughnutChart: React.FC<Props> = ({title, width, height, data}) => {
     const canvasRef = useRef<HTMLCanvasElement | null>(null);
     const chartRef = useRef<Chart<'doughnut'> | null>(null);
+    const prevData = useRef(data);
 
     useEffect(() => {
-        if (!data || !canvasRef.current) {
-            return;
-        }
+        const isEqual = data === prevData.current || deepEqual(prevData.current, data);
+
+        if (!isEqual) {
+            if (chartRef.current) {
+                chartRef.current.destroy();
+            }
 
-        const ctx = canvasRef.current.getContext('2d') as CanvasRenderingContext2D;
-        const dataCopy = JSON.parse(JSON.stringify(data));
+            if (canvasRef.current && data) {
+                const ctx = canvasRef.current.getContext('2d');
+                if (ctx) {
+                    const dataCopy = JSON.parse(JSON.stringify(data));
+                    chartRef.current = new Chart(ctx, {
+                        type: 'doughnut',
+                        data: dataCopy,
+                        options: {},
+                    });
+                }
+            }
 
-        if (chartRef.current) {
-            chartRef.current.update();
-        } else {
-            chartRef.current = new Chart(ctx, {
-                type: 'doughnut',
-                data: dataCopy,
-                options: {},
-            });
+            prevData.current = data;
```

Comment:
Just for clarity, I would move the logic of keeping track of the previous value to another effect. Also, I would put it on the return function of the effect, so it is only change when we are passing to the next render.

Reply:
Done!